### ISSUE TITLE
Deadlock in process shutdown during RCW cleanup

### ISSUE THREAD

**Issue by miguep** [dotnet/coreclr#24125]
We have just recently started to run into this deadlock in coreclr_t::shutdown.

this is what the main thread looks like:
>	ntdll.dll!_KiFastSystemCallRet@0()	Unknown
 	ntdll.dll!_ZwWaitForSingleObject@12()	Unknown
 	KernelBase.dll!WaitForSingleObjectEx(hHandle, dwMilliseconds, bAlertable)	C
 	hostpolicy.dll!coreclr_t::shutdown()	C++
 	hostpolicy.dll!run_as_app()	C++
 	hostpolicy.dll!_corehost_main()	C++
 	hostfxr.dll!fx_muxer_t::read_config_and_execute()	C++
 	hostfxr.dll!fx_muxer_t::handle_exec_host_command()	C++
 	hostfxr.dll!fx_muxer_t::execute()	C++
 	hostfxr.dll!_hostfxr_main_startupinfo()	C++
 	QualityVaultFrontEnd.exe!exe_start()	C++
 	QualityVaultFrontEnd.exe!_wmain()	C++
 	[Inline Frame] QualityVaultFrontEnd.exe!invoke_main()	C++
 	QualityVaultFrontEnd.exe!__scrt_common_main_seh()	C++
 	kernel32.dll!BaseThreadInitThunk(RunProcessInit, StartAddress, Argument)	C
 	ntdll.dll!__RtlUserThreadStart(StartAddress, Argument)	C
 	ntdll.dll!_RtlUserThreadStart(StartAddress, Argument)	C

This only seems to happen when running the master (3.0.100-preview5-011487) bits, the latest release ( 3.0.100-preview4-011228) bits do not repro this issue.

**Comment by jkotas** [comment]
cc @vitek-karas  @sdmaclea @jeffschwMSFT 

Looks like preview4 -> preview5 regression

**Comment by vitek-karas** [comment]
@elinor-fung 
Any chance of getting a dump or something similar, so that we can drill into this a bit more?

**Comment by jkotas** [comment]
@miguep If dump is not possible, stacktraces from other threads would help too. The main thread is waiting for a lock that is held by some other thread.

**Comment by sdmaclea** [comment]
Seems surprising to me that coreclr is not in the trace
```
KernelBase.dll!WaitForSingleObjectEx(hHandle, dwMilliseconds, bAlertable) C
hostpolicy.dll!coreclr_t::shutdown�() C++
```

**Comment by sdmaclea** [comment]
Did preview5 get all the preview4 fixes that went in after we branched?

**Comment by jeffschwMSFT** [comment]
Right now Preview5 is master... Any changes that went into Preview4 need to be manually also merged into master.

**Comment by jkotas** [comment]
> Seems surprising to me that coreclr is not in the trace

Yes, the stacktrace is missing frames. It can be waiting on this lock
https://github.com/dotnet/core-setup/blob/859ee7d256a043b0a9bfe3e9cc6c8483eb23efd5/src/corehost/cli/hostpolicy/coreclr.cpp#L175

**Comment by sdmaclea** [comment]
That lock makes no sense.  Unless multiple threads are calling shutdown...

**Comment by vitek-karas** [comment]
@AaronRobinsonMSFT 
This change has modified the lock in question:
https://github.com/dotnet/core-setup/commit/d0e3248d57e20c6eba223c6ca9e18dafbb9cea72

We don't know for sure it's that lock, but it's likely. I remember that there were some "things" going on with the use of "atomic" specifically around the release branch (the fix was done in the release branch directly). @AaronRobinsonMSFT any ideas?

**Comment by AaronRobinsonMSFT** [comment]
@vitek-karas There was an issue that was addressed around the constructor also calling `shutdown` and that resulted in a lock. In this case, there should only be one place that is calling `shutdown` so I have no idea how this could be a deadlock there. We should make sure the customer has changes post https://github.com/dotnet/core-setup/commit/d18247239a28e4ec353f92bedf2dde153a8068ea

**Comment by elinor-fung** [comment]
@miguep provided a repro machine. It does not look to be from that lock. I just started looking into it: we're getting into `coreclr_shutdown_2` and blocking waiting for the finalizer thread, which appears to be cleaning up RCWs.

**Comment by AaronRobinsonMSFT** [comment]
Also all atomic issues have been resolved now that we statically link against atomic.a.

**Comment by AaronRobinsonMSFT** [comment]
@elinor-fung has a dump of the issue. This seems to be some issue with RCW cleanup. I will have to look into it.

**Comment by AaronRobinsonMSFT** [comment]
Okay, this is definitely in coreclr binary and not in any not core-setup code.

main thread
```cpp
> 	coreclr.dll!CLREventWaitHelper(void * handle, unsigned long dwMilliseconds, int alertable) Line 402	C++
 	coreclr.dll!CLREventBase::WaitEx(unsigned long dwMilliseconds, WaitMode mode, PendingSync * syncState) Line 471	C++
 	coreclr.dll!CLREventBase::Wait(unsigned long dwMilliseconds, int alertable, PendingSync * syncState) Line 418	C++
 	coreclr.dll!FinalizerThread::RaiseShutdownEvents() Line 72	C++
 	coreclr.dll!EEShutDownHelper(int fIsDllUnloading) Line 1533	C++
 	coreclr.dll!EEShutDown(int fIsDllUnloading) Line 1939	C++
 	coreclr.dll!CorHost2::UnloadAppDomain2(unsigned long dwDomainId, int fWaitUntilDone, int * pLatchedExitCode) Line 1048	C++
	coreclr.dll!coreclr_shutdown_2(void * hostHandle, unsigned int domainId, int * latchedExitCode) Line 323	C++
 	hostpolicy.dll!coreclr_t::shutdown(int *)	C++
 	hostpolicy.dll!run_as_app(class std::shared_ptr<class coreclr_t> const &,struct hostpolicy_context_t const &,int,wchar_t const * *)	C++
 	hostpolicy.dll!_corehost_main()	C++
 	hostfxr.dll!fx_muxer_t::read_config_and_execute(class std::basic_string<wchar_t,struct std::char_traits<wchar_t>,class std::allocator<wchar_t> > const &,struct host_startup_info_t const &,class std::basic_string<wchar_t,struct std::char_traits<wchar_t>,class std::allocator<wchar_t> > const &,class std::unordered_map<class std::basic_string<wchar_t,struct std::char_traits<wchar_t>,class std::allocator<wchar_t> >,class std::vector<class std::basic_string<wchar_t,struct std::char_traits<wchar_t>,class std::allocator<wchar_t> >,class std::allocator<class std::basic_string<wchar_t,struct std::char_traits<wchar_t>,class std::allocator<wchar_t> > > >,struct std::hash<class std::basic_string<wchar_t,struct std::char_traits<wchar_t>,class std::allocator<wchar_t> > >,struct std::equal_to<class std::basic_string<wchar_t,struct std::char_traits<wchar_t>,class std::allocator<wchar_t> > >,class std::allocator<struct std::pair<class std::basic_string<wchar_t,struct std::char_traits<wchar_t>,class std::allocator<wchar_t> > const ,class	C++
 	hostfxr.dll!fx_muxer_t::handle_exec_host_command(class std::basic_string<wchar_t,struct std::char_traits<wchar_t>,class std::allocator<wchar_t> > const &,struct host_startup_info_t const &,class std::basic_string<wchar_t,struct std::char_traits<wchar_t>,class std::allocator<wchar_t> > const &,class std::unordered_map<class std::basic_string<wchar_t,struct std::char_traits<wchar_t>,class std::allocator<wchar_t> >,class std::vector<class std::basic_string<wchar_t,struct std::char_traits<wchar_t>,class std::allocator<wchar_t> >,class std::allocator<class std::basic_string<wchar_t,struct std::char_traits<wchar_t>,class std::allocator<wchar_t> > > >,struct std::hash<class std::basic_string<wchar_t,struct std::char_traits<wchar_t>,class std::allocator<wchar_t> > >,struct std::equal_to<class std::basic_string<wchar_t,struct std::char_traits<wchar_t>,class std::allocator<wchar_t> > >,class std::allocator<struct std::pair<class std::basic_string<wchar_t,struct std::char_traits<wchar_t>,class std::allocator<wchar_t> > const ,clas	C++
 	hostfxr.dll!fx_muxer_t::execute(class std::basic_string<wchar_t,struct std::char_traits<wchar_t>,class std::allocator<wchar_t> >,int,wchar_t const * * const,struct host_startup_info_t const &,wchar_t * const,int,int *)	C++
 	hostfxr.dll!_hostfxr_main_startupinfo()	C++
```

Clean up
```c++
> 	coreclr.dll!CtxEntry::EnterContext(HRESULT(__stdcall*)(void *) pCallbackFunc, void * pData) Line 1511	C++
 	coreclr.dll!RCW::EnterContext(HRESULT(__stdcall*)(void *) pCallbackFunc, void * pData) Line 4204	C++
 	coreclr.dll!RCWCleanupList::ReleaseRCWListInCorrectCtx(void * pData) Line 1862	C++
 	coreclr.dll!RCWCleanupList::CleanupAllWrappers() Line 1690	C++
 	coreclr.dll!SyncBlockCache::CleanupSyncBlocks() Line 626	C++
	coreclr.dll!Thread::DoExtraWorkForFinalizer() Line 7315	C++
 	coreclr.dll!FinalizerThread::FinalizerThreadWorker(void * args) Line 494	C++
 	[Inline Frame] coreclr.dll!ManagedThreadBase_DispatchInner(ManagedThreadCallState *) Line 7394	C++
 	coreclr.dll!ManagedThreadBase_DispatchMiddle(ManagedThreadCallState * pCallState) Line 7438	C++
 	coreclr.dll!ManagedThreadBase_DispatchOuter(ManagedThreadCallState * pCallState) Line 7624	C++
 	coreclr.dll!ManagedThreadBase_NoADTransition(void(__stdcall*)(void *)) Line 7675	C++
 	coreclr.dll!FinalizerThread::FinalizerThreadStart(void * args) Line 590	C++
 	coreclr.dll!Thread::intermediateThreadProc(void * arg) Line 2099	C++
```

**Comment by AaronRobinsonMSFT** [comment]
Nevermind. 🙄

**Comment by jkotas** [comment]
This may be related to my change https://github.com/dotnet/coreclr/pull/23595/files#diff-7817a3c11b2998ad499862a6f4106323L843 . Before this change, this wait had 2 second timeout. We may add the 2 second timeout back, or just stop cleaning up the RCWs.

**Comment by AaronRobinsonMSFT** [comment]
@jkotas Very sad panda. That looks like the issue. Any strong opinions on which direction? I am thinking of skipping RCW cleanup during shutdown rather than reinserting the timeout.

**Comment by jkotas** [comment]
I am worried that we may need to add the timeout back. The finalizer may be in the middle of regular RCW cleanup when the shutdown happens. If the RCW cleanup can hang during shutdown, we would still get intermittent hang in this case.

Do you know why `EnterContext` is blocked in the dump you have looked at? What would need to happen to make it finish gracefully?

**Comment by jkotas** [comment]
Maybe the fix should be to change https://github.com/dotnet/coreclr/blob/58d9cf157f54e8fd61eaaf56b3f8045075d171cd/src/vm/finalizerthread.h#L71 to pass `alertable=TRUE` ?
